File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportC#/Persistence/ado/TripDbRepository.cs
using FirmaTransportC_.Model;
using FirmaTransportC_.Persistence.utils;
using FirmaTransportC_.Persistence;
using log4net;
using System;
using System.Collections.Generic;
using System.Data;
using System.Globalization;

namespace FirmaTransportC_.Persistence.ado
{
    public class TripDbRepository : ITripRepository
    {
        private static readonly ILog log = LogManager.GetLogger("TripDbRepository");
        private readonly IDictionary<string, string> props;

        public TripDbRepository(IDictionary<string, string> props)
        {
            log.Info("Initializing TripDbRepository with properties: " + string.Join(", ", props));
            this.props = props;
        }

        public long Add(Trip entity)
        {
            log.InfoFormat("Adding trip: {0}", entity);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = @"
                    INSERT INTO trips (destination, date, hour, available_seats) 
                    VALUES (@destination, @date, @hour, @seats);
                    SELECT last_insert_rowid();
                ";
                command.Parameters.Add(CreateParam(command, "@destination", entity.Destination));
                command.Parameters.Add(CreateParam(command, "@date", entity.Date));
                command.Parameters.Add(CreateParam(command, "@hour", entity.Hour));
                command.Parameters.Add(CreateParam(command, "@seats", entity.AvailableSeats));
                var result = command.ExecuteScalar();
                if (result != null)
                {
                    long id = Convert.ToInt64(result);
                    entity.Id = id;
                    return id;
                }
            }
            catch (Exception ex)
            {
                log.Error("Error adding trip", ex);
            }
            return -1;
        }

        public void Delete(long id)
        {
            log.InfoFormat("Deleting trip with id: {0}", id);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "DELETE FROM trips WHERE id = @id";
                command.Parameters.Add(CreateParam(command, "@id", id));
                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error deleting trip", ex);
            }
        }

        public void Update(Trip entity)
        {
            log.InfoFormat("Updating trip: {0}", entity);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = @"
                    UPDATE trips 
                    SET destination = @destination, date = @date, hour = @hour, available_seats = @seats 
                    WHERE id = @id
                ";
                command.Parameters.Add(CreateParam(command, "@destination", entity.Destination));
                command.Parameters.Add(CreateParam(command, "@date", entity.Date));
                command.Parameters.Add(CreateParam(command, "@hour", entity.Hour));
                command.Parameters.Add(CreateParam(command, "@seats", entity.AvailableSeats));
                command.Parameters.Add(CreateParam(command, "@id", entity.Id));
                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                log.Error("Error updating trip", ex);
            }
        }

        public Trip FindOne(long id)
        {
            log.InfoFormat("Finding trip by id: {0}", id);
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "SELECT * FROM trips WHERE id = @id";
                command.Parameters.Add(CreateParam(command, "@id", id));
                using var reader = command.ExecuteReader();
                if (reader.Read())
                    return ExtractTrip(reader);
            }
            catch (Exception ex)
            {
                log.Error("Error finding trip", ex);
            }
            return null;
        }

        public List<Trip> FindAll()
        {
            log.Info("Finding all trips");
            var trips = new List<Trip>();
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = "SELECT * FROM trips";
                using var reader = command.ExecuteReader();
                while (reader.Read())
                    trips.Add(ExtractTrip(reader));
            }
            catch (Exception ex)
            {
                log.Error("Error finding trips", ex);
            }
            return trips;
        }

        public List<Trip> FindTripsByDestinationAndDepartureDate(string destination, string date, string hour)
        {
            log.InfoFormat("Finding trips by destination: {0}, hour: {1} and date: {2}", destination, hour, date);
            var trips = new List<Trip>();
            using var connection = DbUtils.GetConnection(props);
            try
            {
                using var command = connection.CreateCommand();
                command.CommandText = @"
                    SELECT * FROM trips
                    WHERE destination = @destination
                    AND date = @date
                    AND hour = @hour
                ";
                command.Parameters.Add(CreateParam(command, "@destination", destination));
                command.Parameters.Add(CreateParam(command, "@date", date));
                command.Parameters.Add(CreateParam(command, "@hour", hour));
                using var reader = command.ExecuteReader();
                while (reader.Read())
                    trips.Add(ExtractTrip(reader));
            }
            catch (Exception ex)
            {
                log.Error("Error finding filtered trips", ex);
            }
            return trips;
        }

        private Trip ExtractTrip(IDataReader reader)
        {
            long id = reader.GetInt64(0);
            string destination = reader.GetString(1);
            string date = reader.GetString(2);
            int availableSeats = reader.GetInt32(3);
            string hour = reader.GetString(4);
            return new Trip(id, destination, date, hour, availableSeats);
        }

        private IDbDataParameter CreateParam(IDbCommand cmd, string name, object value)
        {
            var p = cmd.CreateParameter();
            p.ParameterName = name;
            p.Value = value;
            return p;
        }
    }
}

File: Semestrul 4/Medii de proiectare si programare/Laboratoare/FirmaTransportJava/Persistence/src/main/java/transport/persistence/repository/jdbc/TripDbRepository.java
package transport.persistence.repository.jdbc;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import transport.model.Trip;
import transport.persistence.repository.TripRepository;
import transport.persistence.repository.utils.JdbcUtils;

import java.sql.*;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class TripDbRepository implements TripRepository {

    private final JdbcUtils dbUtils;
    private static final Logger logger = LogManager.getLogger();

    public TripDbRepository(Properties properties) {
        logger.info("Initializing TripDbRepository with properties: {}", properties);
        dbUtils = new JdbcUtils(properties);
    }

    @Override
    public Long add(Trip entity) {
        logger.info("Adding trip: {}", entity);

        String sql = "INSERT INTO trips (destination, date, hour, available_seats) VALUES (?, ?, ?, ?)";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql, Statement.RETURN_GENERATED_KEYS)) {

            statement.setString(1, entity.getDestination());
            statement.setString(2, entity.getDate());
            statement.setString(3, entity.getHour());
            statement.setInt(4, entity.getAvailableSeats());

            int result = statement.executeUpdate();
            logger.trace("Inserted rows: {}", result);

            try (ResultSet keys = statement.getGeneratedKeys()) {
                if (keys.next()) {
                    long id = keys.getLong(1);
                    entity.setId(id);
                    return id;
                }
            }

        } catch (SQLException e) {
            logger.error("Error adding trip", e);
            System.out.println("Error adding trip: " + e);
        }

        return null;
    }

    @Override
    public void delete(Long id) {
        logger.info("Deleting trip with id: {}", id);

        String sql = "DELETE FROM trips WHERE id = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, id);
            statement.executeUpdate();

        } catch (SQLException e) {
            logger.error("Error deleting trip", e);
            System.out.println("Error deleting trip: " + e);
        }
    }

    @Override
    public void update(Trip entity) {
        logger.info("Updating trip: {}", entity);

        String sql = "UPDATE trips SET destination = ?, date = ?, hour = ?, available_seats = ? WHERE id = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setString(1, entity.getDestination());
            statement.setString(2, entity.getDate());
            statement.setString(3, entity.getHour());
            statement.setInt(4, entity.getAvailableSeats());
            statement.setLong(5, entity.getId());

            statement.executeUpdate();

        } catch (SQLException e) {
            logger.error("Error updating trip", e);
            System.out.println("Error updating trip: " + e);
        }
    }

    @Override
    public Trip findById(Long id) {
        logger.info("Finding trip by id: {}", id);

        String sql = "SELECT * FROM trips WHERE id = ?";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setLong(1, id);

            try (ResultSet rs = statement.executeQuery()) {
                if (rs.next()) {
                    return extractTrip(rs);
                }
            }

        } catch (SQLException e) {
            logger.error("Error finding trip", e);
            System.out.println("Error finding trip: " + e);
        }

        return null;
    }

    @Override
    public List<Trip> findAll() {
        logger.info("Finding all trips");

        List<Trip> trips = new ArrayList<>();
        String sql = "SELECT * FROM trips";

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql);
             ResultSet rs = statement.executeQuery()) {

            while (rs.next()) {
                trips.add(extractTrip(rs));
            }

        } catch (SQLException e) {
            logger.error("Error finding trips", e);
            System.out.println("Error finding trips: " + e);
        }

        return trips;
    }

    @Override
    public List<Trip> findTripsByDestinationAndDepartureDate(String destination, String date, String hour) {
        logger.info("Finding trips by destination: {} , date: {} and hour: {}", destination, date, hour);

        List<Trip> trips = new ArrayList<>();

        String sql = """
                SELECT * FROM trips
                WHERE destination = ?
                AND date = ?
                AND hour = ?
                """;

        try (Connection connection = dbUtils.getConnection();
             PreparedStatement statement = connection.prepareStatement(sql)) {

            statement.setString(1, destination);
            statement.setString(2, date);
            statement.setString(3, hour);

            try (ResultSet rs = statement.executeQuery()) {
                while (rs.next()) {
                    trips.add(extractTrip(rs));
                }
            }

        } catch (SQLException e) {
            logger.error("Error finding filtered trips", e);
            System.out.println("Error finding trips: " + e);
        }

        return trips;
    }

    private Trip extractTrip(ResultSet rs) throws SQLException {
        Long id = rs.getLong("id");
        String destination = rs.getString("destination");
        String date = rs.getString("date");
        String hour = rs.getString("hour");
        int availableSeats = rs.getInt("available_seats");

        return new Trip(id, destination, date, hour, availableSeats);
    }
}